### ISSUE TITLE
_children() called on directory entries during breadth-first traversal

### ISSUE THREAD

**Issue by djerius** [dagolden/Path-Iterator-Rule#20]
Given a directory structure:

    p
    ├── p/a
    │   ├── p/a/1
    │   ├── p/a/2
    │   └── p/a/3
    ├── p/b
    │   ├── p/b/4
    │   ├── p/b/5
    │   └── p/b/6
    └── p/c
        ├── p/c/7
        ├── p/c/8
        └── p/c/9

I'd expect the following order of operations for a breadth-first
traversal

1. visit( $_ ) for readdir( p )
2. visit( $_ ) for readdir( p/a )
3. visit( $_ ) for readdir( p/b )
4. visit( $_ ) for readdir( p/c )

Instead, the order of operation is

1. readdir ( p )
2. visit ( p/a )
3. readdir ( p/a )
4. visit ( p/b )
5. readdir ( p/b )
6. visit ( p/c )
7. readdir ( p/c )
6. visit( $_ ) for entries in (p/a)
6. visit( $_ ) for entries in (p/b)
6. visit( $_ ) for entries in (p/c)

The [attached code ](https://github.com/dagolden/Path-Iterator-Rule/files/974298/foo.txt) illustrates this behavior:

    % perl foo.pl
    .
    CHILDREN: .
    ./p
    CHILDREN: ./p
    ./p/a
    CHILDREN: ./p/a
    ./p/b
    CHILDREN: ./p/b
    ./p/c
    CHILDREN: ./p/c
    ./p/a/1
    ./p/a/2
    ./p/a/3
    ./p/b/4
    ./p/b/5
    ./p/b/6
    ./p/c/7
    ./p/c/8
    ./p/c/9

The disadvantage of this approach is that the children are determined before it is known if they are actually needed (e.g. if the traversal is aborted) and it may be expensive to perform a directory read (e.g. if it is on a sshfs mounted disk, on an FTP site, etc).

I don't have any tuits at the moment, but if I did, I'd wrap the _children calls in a closure, insert them into the queue, and modify the queue popping code to push the results of any closures on the queue back onto it.

Thanks,

Diab

**Comment by djerius** [comment]
Here's a slightly changed version of the [test script](https://github.com/dagolden/Path-Iterator-Rule/files/1699904/depth.txt) which makes the output a bit easier to understand.

Here's what the existing PIR code outputs:
```
VISIT: .
CHILDREN: .
VISIT: ./p
CHILDREN: ./p
VISIT: ./p/a
CHILDREN: ./p/a
VISIT: ./p/b
CHILDREN: ./p/b
VISIT: ./p/c
CHILDREN: ./p/c
VISIT: ./p/a/1
VISIT: ./p/a/2
VISIT: ./p/a/3
VISIT: ./p/b/4
VISIT: ./p/b/5
VISIT: ./p/b/6
VISIT: ./p/c/7
VISIT: ./p/c/8
VISIT: ./p/c/9
```
And here's what I expect for a breadth-first traversal:
```
VISIT: .
CHILDREN: .
VISIT: ./p
CHILDREN: ./p
VISIT: ./p/a
VISIT: ./p/b
VISIT: ./p/c
CHILDREN: ./p/a
VISIT: ./p/a/1
VISIT: ./p/a/2
VISIT: ./p/a/3
CHILDREN: ./p/b
VISIT: ./p/b/4
VISIT: ./p/b/5
VISIT: ./p/b/6
CHILDREN: ./p/c
VISIT: ./p/c/7
VISIT: ./p/c/8
VISIT: ./p/c/9
```

To clarify why this makes a difference, suppose the search should be terminated early because of some business logic:

```
my $next = $rule->iter( @dirs, \%options);
while ( defined( my $file = $next->() ) ) {
  last if ...;  # abort search
}
```
at `./p/a/3`.  In the existing code,  directories `./p/b` and `./p/c` have already been scanned before PIR visits the entries in `./p/a`, but that is unnecessary and in the case of remote or large filesystems can be very expensive.

**Comment by xdg** [comment]
Fixed via #22
